### PR TITLE
Example ERC20 contract in Yul.

### PR DESCRIPTION
**Pull request by chriseth**
References #8523

### CODE REVIEW

**Comment by chriseth** [comment]
The example still needs to be linked from one of the earlier sections, and it needs to be checked for correctness. I would be glad if someone else could take that.

Just for those curious, this is the optimized output:

```
object "Token" {
    code {
        {
            sstore(0, caller())
            let _1 := datasize("runtime")
            datacopy(0, dataoffset("runtime"), _1)
            return(0, _1)
        }
    }
    object "runtime" {
        code {
            {
                if iszero(iszero(callvalue())) { revert(0, 0) }
                let _1 := 0
                switch div(calldataload(_1), shl(224, 1))
                case 0x70a08231 {
                    let v := decodeAsUint(_1)
                    if iszero(iszero(and(v, shl(160, 0xffffffffffffffffffffffff)))) { revert(_1, _1) }
                    returnUint(sload(add(0x1000, v)))
                }
                case 0x18160ddd { returnUint(sload(1)) }
                case 0xa9059cbb {
                    transfer(decodeAsAddress(_1), decodeAsUint(1))
                }
                case 0x40c10f19 {
                    mint(decodeAsAddress(_1), decodeAsUint(1))
                }
                default { revert(_1, _1) }
            }
            function mint(account, amount)
            {
                if iszero(eq(sload(0), caller())) { revert(0, 0) }
                sstore(1, safeAdd(sload(1), amount))
                let offset := add(0x1000, account)
                sstore(offset, safeAdd(sload(offset), amount))
            }
            function transfer(to, amount)
            {
                let _1 := 0x1000
                let offset := add(_1, caller())
                let bal := sload(offset)
                if lt(bal, amount) { revert(0, 0) }
                sstore(offset, sub(bal, amount))
                let offset_1 := add(_1, to)
                sstore(offset_1, safeAdd(sload(offset_1), amount))
            }
            function decodeAsAddress(offset) -> v
            {
                v := decodeAsUint(offset)
                if iszero(iszero(and(v, shl(160, 0xffffffffffffffffffffffff)))) { revert(0, 0) }
            }
            function decodeAsUint(offset) -> v
            {
                let _1 := mul(offset, 0x20)
                if lt(calldatasize(), add(_1, 36)) { revert(v, v) }
                v := calldataload(add(4, _1))
            }
            function returnUint(v)
            {
                mstore(0, v)
                return(0, 0x20)
            }
            function safeAdd(a, b) -> r
            {
                r := add(a, b)
                if or(lt(r, a), lt(r, b)) { revert(0, 0) }
            }
        }
    }
}

```

File: docs/yul.rst [hunk 1]
```diff
@@ -1015,3 +1015,117 @@ If you want to use Solidity in stand-alone Yul mode, you activate the optimizer
...
+                    transfer(decodeAsAddress(0), decodeAsUint(1))
+                }
+                case 0x40c10f19 /* "mint(address,uint256)" */ {
+                    mint(decodeAsAddress(0), decodeAsUint(1))
+                }
+                default {
+                    revert(0, 0)
+                }
+
+                function mint(account, amount) {
```
Comment (hrkrshnn) [inline docs/yul.rst:1146]: The functions mint, transfer, totalSupply, balanceOf and their internal calls look good to me. However, the address operations need to be examined by someone who know YUL better.

What about the rest of the standard functions, i.e., transferFrom, approve and allowance?

Comment (chriseth) [reply docs/yul.rst:1146]: As far as I know, transferFrom, approve and allowance are optional in ERC20, and I think we should keep this example minimal.

Comment (hrkrshnn) [reply docs/yul.rst:1146]: According to https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20.md, they (transferFrom, approve and allowance) are not optional.

Comment (chriseth) [reply docs/yul.rst:1146]: You are right - sorry about that! If we want to implement it, we need another data structure and we will have to apply a hash.

Comment (axic) [reply docs/yul.rst:1146]: You can also rename it to "BasicToken" to get over the issue.

**Comment by chriseth** [comment]
This is complete now.

File: docs/yul.rst [hunk 2]
```diff
@@ -1015,3 +1017,198 @@ If you want to use Solidity in stand-alone Yul mode, you activate the optimizer
...
+
+Complete ERC20 Example
+======================
+
+.. code-block:: yul
+
+    object "Token" {
+        code {
+            // Store the creator in slot zero.
+            sstore(0, caller())
```
Comment (leonardoalt) [inline docs/yul.rst:1104]:
```suggestion
            sstore(ownerPos(), caller())
```

File: docs/yul.rst [hunk 3]
```diff
@@ -1015,3 +1017,198 @@ If you want to use Solidity in stand-alone Yul mode, you activate the optimizer
...
+                case 0xa9059cbb /* "transfer(address,uint256)" */ {
+                    transfer(decodeAsAddress(0), decodeAsUint(1))
+                    returnTrue()
+                }
+                case 0x23b872dd /* "transferFrom(address,address,uint256)" */ {
+                    transferFrom(decodeAsAddress(0), decodeAsAddress(1), decodeAsUint(2))
+                    returnTrue()
+                }
+                case 0x095ea7b3 /* "approve(address,uint256)" */ {
+                    approve(decodeAsAddress(0), decodeAsUint(2))
```
Comment (leonardoalt) [inline docs/yul.rst]: Why 2?

File: docs/yul.rst [hunk 4]
```diff
@@ -1015,3 +1017,198 @@ If you want to use Solidity in stand-alone Yul mode, you activate the optimizer
...
+                function emitEvent(signatureHash, indexed1, indexed2, nonIndexed) {
+                    mstore(0, nonIndexed)
+                    log3(0, 0x20, signatureHash, indexed1, indexed2)
+                }
+
+                /* -------- storage layout ---------- */
+                function ownerPos() -> p { p := 0 }
+                function totalSupplyPos() -> p { p := 1 }
+                function accountToStorageOffset(account) -> offset {
+                    offset := add(0x1000, account)
```
Comment (leonardoalt) [inline docs/yul.rst:1219]: Why add 1000?

File: docs/yul.rst (hunk 2 above)
Comment (chriseth) [reply docs/yul.rst:1104]: Ah no! The function is not defined in the creation context!

File: docs/yul.rst (hunk 3 above)
Comment (chriseth) [reply docs/yul.rst]: because bug

File: docs/yul.rst (hunk 4 above)
Comment (chriseth) [reply docs/yul.rst:1219]: we could also add `totalSupplyPos + 1`, but I thought a little space could not hurt.

File: docs/yul.rst (hunk 2 above)
Comment (leonardoalt) [reply docs/yul.rst:1104]: Ah true, sorry!

File: docs/yul.rst (hunk 4 above)
Comment (leonardoalt) [reply docs/yul.rst:1219]: ok, sure

File: docs/yul.rst (hunk 3 above)
Comment (ekpyron) [reply docs/yul.rst]: Still?

Comment (chriseth) [reply docs/yul.rst]: Sorry!

**Review by leonardoalt — APPROVED** [review]
I haven't actually deployed/ran it though.
I'll do that later at some point

**Pull request merged**